core-bundle/src/Resources/contao/library/Contao/Search.php

```diff
@@ -403,116 +403,186 @@ public static function searchFor($strKeywords, $blnOrSearch=false, $arrPid=array
+
+		// Get wildcards
+		foreach ($arrWildcards as $strKeyword)
 		{
-			$strQuery .= ", (SELECT COUNT(*) FROM tl_search_index WHERE (" . implode(' OR ', array_fill(0, $intWildcards, 'word LIKE ?')) . ") AND pid=sid) AS wildcards";
-			$arrValues = array_merge($arrValues, $arrWildcards);
+			$arrMatches[] = \count($arrAllKeywords);
+			$arrAllKeywords[] = 'word LIKE ?';
 		}
 
-		// Count the number of matches
-		$strQuery .= ", COUNT(*) AS count";
-
-		// Get the relevance
-		$strQuery .= ", SUM(relevance) AS relevance";
-
-		// Prepare keywords array
-		$arrAllKeywords = array();
+		// Wildcard values are required three times in the query
+		for ($i = 0; $i < 3; $i++)
+		{
+			foreach ($arrWildcards as $strKeyword)
+			{
+				$arrValues[] = $strKeyword;
+			}
+		}
 
 		// Get keywords
-		if (!empty($arrKeywords))
+		foreach ($arrKeywords as $strKeyword)
 		{
-			$arrAllKeywords[] = implode(' OR ', array_fill(0, \count($arrKeywords), 'word=?'));
-			$arrValues = array_merge($arrValues, $arrKeywords);
-			$intKeywords += \count($arrKeywords);
+			$arrMatches[] = \count($arrAllKeywords);
+			$arrAllKeywords[] = 'word=?';
+			$arrValues[] = $strKeyword;
 		}
 
 		// Get included keywords
-		if ($intIncluded)
+		foreach ($arrIncluded as $strKeyword)
+		{
+			$arrRequiredMatches[] = \count($arrAllKeywords);
+			$arrAllKeywords[] = 'word=?';
+			$arrValues[] = $strKeyword;
+		}
+
+		// Get excluded keywords
+		foreach ($arrExcluded as $strKeyword)
 		{
-			$arrAllKeywords[] = implode(' OR ', array_fill(0, $intIncluded, 'word=?'));
-			$arrValues = array_merge($arrValues, $arrIncluded);
-			$intKeywords += $intIncluded;
+			$arrExcludedMatches[] = \count($arrAllKeywords);
+			$arrAllKeywords[] = 'word=?';
+			$arrValues[] = $strKeyword;
 		}
 
 		// Get keywords from phrases
-		if ($intPhrases)
+		foreach ($arrPhrases as $strPhrase)
 		{
-			foreach ($arrPhrases as $strPhrase)
+			foreach (self::splitIntoWords(str_replace('[^[:alnum:]]+', ' ', $strPhrase), $GLOBALS['TL_LANGUAGE']) as $strKeyword)
 			{
-				$arrWords = self::splitIntoWords(str_replace('[^[:alnum:]]+', ' ', $strPhrase), $GLOBALS['TL_LANGUAGE']);
-				$arrAllKeywords[] = implode(' OR ', array_fill(0, \count($arrWords), 'word=?'));
-				$arrValues = array_merge($arrValues, $arrWords);
-				$intKeywords += \count($arrWords);
+				$arrMatches[] = \count($arrAllKeywords);
+				$arrAllKeywords[] = 'word=?';
+				$arrValues[] = $strKeyword;
 			}
 		}
 
-		// Get wildcards
-		if ($intWildcards)
+		// Get the relevance
+		$strQuery .= ", (0";
+
+		foreach ($arrAllKeywords as $index => $strKeywordExpression)
 		{
-			$arrAllKeywords[] = implode(' OR ', array_fill(0, $intWildcards, 'word LIKE ?'));
-			$arrValues = array_merge($arrValues, $arrWildcards);
+			if (\in_array($index, $arrExcludedMatches, true))
+			{
+				continue;
+			}
+
+			if (isset($arrWildcards[$index]))
+			{
+				$strQuery .= "+ (
+					(1+LOG(SUM(match$index * relevance))) * POW(LOG((
+						SELECT COUNT(*) FROM tl_search
```

Comment:
Can't we do the `COUNT` query first as a separate query and use the result as a constant? Not sure but could be faster imho :)

Reply:
Yes, that would be faster. But not all of our supported MySQL versions support CTEs (common table expressions) and therefore this cannot be done in one single query. I tried using MySQL variables to optimize this but they are kind of buggy and don’t seem to work reliably.

A temporary table might help, but I’m not sure about the overhead this would cost.

Reply:
Using variables in b9db1675ecb06980096985477c07866795f22900 now.